backend/applications.go

```diff
@@ -0,0 +1,257 @@
+package main
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/base64"
+	"net/http"
+	"strings"
+
+	"github.com/elonsoc/center/backend/service"
+	"github.com/go-chi/chi/v5"
+	_ "github.com/lib/pq"
+)
+
+/*
+	This file deals with application registration and everything that comes with it.
+	- recieving the registration form
+	- generating a project ID
+	- generating an API key
+	- storing the information in the database
+	- returning the pertinent information to the user
+
+	Database Design can be found: https://docs.google.com/document/d/1zEK9K7crTcCcE9bMKm87qRHCyqa5I0_vjuI9eqYSaLg/edit?usp=sharing
+*/
+
+// Defining the struct for the applications router
+type ApplicationsRouter struct {
+	chi.Router
+	Svcs *service.Service
+}
+
+// NewApplicationsRouter is a function that returns a new applications router
+func NewApplicationsRouter(a *ApplicationsRouter) *ApplicationsRouter {
+	a.Svcs.Logger.Info("Initializing applications router")
+
+	r := chi.NewRouter()
+	r.Post("/", a.newApp)
+	r.Get("/", a.myApps)
+
+	a.Router = r
+	a.Svcs.Logger.Info("Applications router initialized")
+	return a
+}
+
+// create a registration structure (This is not entirely necessary, but it makes things easier)
+// Contents of this might change as the project progresses
+type application struct {
+	appName     string
+	appID       string
+	description string
+	owners      string
+	teamName    string
+	apiKey      string
+	isValid     bool
+}
+
+// This function handles the registration form. It is called when the user submits a registration form.
+// It will parse the form, generate a project ID and API key, and store the information in the database.
+// It will then return the pertinent information to the user.
+func (ar *ApplicationsRouter) newApp(w http.ResponseWriter, r *http.Request) {
+	ctx := context.Background()
+	// Create a new Application struct
+	app := application{}
+	// parse the request body into the registration variable
+	app.apiKey = ar.apiKeyGenerate()
+	app.appName = r.FormValue("title")
+	app.description = r.FormValue("description")
+	app.owners = r.FormValue("owners")
+	app.teamName = r.FormValue("teamName")
+	app.appID = ar.appIDGenerate()
+	app.isValid = true
+
+	/* Storing information in the Database*/
+
+	// Initiating a transaction
+	tx, err := ar.Svcs.Db.Begin(ctx)
+	if err != nil {
+		ar.Svcs.Logger.Error(err)
+		ar.Svcs.Logger.Info("Error initiating database transaction")
+		return
+	}
+	defer tx.Rollback(ctx)
+
+	// Getting prepared statements ready to use later
+	keyStmt, err := tx.Prepare(ctx, "insert_into_keys", "INSERT INTO keys (id, apiKey, isValid) VALUES ($1, $2, $3)")
```

Comment:
We might want to move these Prepare statements into [services.go](https://github.com/elonsoc/ods/blob/390054fdfc361f94d74fc98c753bb862cc3bf997/backend/service/services.go#L67-L68) where we define the Db connection and then referring to it in the function:

in `services.go`
```go
func InitDB(databaseURL string, log *logrus.Logger) *pgx.Conn {
	...
	connection.Prepare(context.Background(), "insert_into_keys", "INSERT INTO keys (id, apiKey, isValid) VALUES ($1, $2, $3)")
	.
	.
	return connection
}
```
```suggestion
	keyStmt, err := tx.Exec(ctx, "insert_into_keys")
```

Reply:
I have just done this. I prepared 2 statements in services.go. I only prepared the 2 statements that deal with user input. This is because the other queries are base62 and wont DROP ALL TABLES. Also pgx v5 has built in prepared statements after first execution, and that should be more than fine for those queries.